xde/src/xde.rs

```diff
@@ -298,16 +326,119 @@ unsafe extern "C" fn _fini() -> c_int {
 }
 
 #[no_mangle]
-unsafe extern "C" fn xde_ioctl(
-    _dev: dev_t,
-    _cmd: c_int,
-    _arg: intptr_t,
-    _mode: c_int,
+unsafe extern "C" fn xde_qopen(
+    q: *mut queue_t,
+    devp: *mut dev_t,
+    oflag: c_int,
+    sflag: c_int,
+    credp: *mut cred_t,
+) -> c_int {
+    assert!(!xde_dip.is_null());
+
+    // open called on a non-MAC managed minor, treat as character device.
+    // See comment in `_init`.
+    if getminor(*devp) >= mac_private_minor() {
+        return ENOSTR;
+    }
+
+    // Safety: initialized in `_init`.
+    let orig_open = (*(*xde_cb_ops.cb_str).st_wrinit).qi_qopen;
+    orig_open(q, devp, oflag, sflag, credp)
+}
+
+#[no_mangle]
+unsafe extern "C" fn xde_open(
+    devp: *mut dev_t,
+    _flag: c_int,
+    otyp: c_int,
     _credp: *mut cred_t,
-    _rvalp: *mut c_int,
 ) -> c_int {
-    warn!("xde_ioctl not supported, use dld_ioc");
-    ENOTSUP
+    assert!(!xde_dip.is_null());
+
+    if otyp != OTYP_CHR {
+        return EINVAL;
+    }
+
+    let minor = getminor(*devp);
+    if minor != XDE_CTL_MINOR {
+        warn!("invalid minor number for open: {minor}");
+        return ENXIO;
+    }
+
+    // TODO: check flags
+
+    // TODO: device cloning needed?
+    *devp = makedevice(getmajor(*devp), minor);
+
+    0
+}
+
+#[no_mangle]
+unsafe extern "C" fn xde_close(
+    dev: dev_t,
+    _flag: c_int,
+    otyp: c_int,
+    _credp: *mut cred_t,
+) -> c_int {
+    assert!(!xde_dip.is_null());
+
+    if otyp != OTYP_CHR {
+        return EINVAL;
+    }
+
+    let minor = getminor(dev);
+    if minor != XDE_CTL_MINOR {
+        warn!("invalid minor number for close: {minor}");
+        return ENXIO;
+    }
+
+    0
+}
+
+#[no_mangle]
+unsafe extern "C" fn xde_ioctl(
+    dev: dev_t,
+    cmd: c_int,
+    arg: intptr_t,
+    mode: c_int,
+    credp: *mut cred_t,
+    rvalp: *mut c_int,
+) -> c_int {
+    assert!(!xde_dip.is_null());
+
+    let minor = getminor(dev);
+    if minor != XDE_CTL_MINOR {
+        warn!("invalid minor number for ioctl: {minor}");
+        return EINVAL;
+    }
+
+    if cmd != XDE_IOC_OPTE_CMD {
+        warn!("invalid ioctl command: {cmd}");
+        return EINVAL;
+    }
+
+    match secpolicy::secpolicy_dl_config(credp) {
+        0 => {}
+        err => {
+            warn!("secpolicy_dl_config failed: {err}");
+            return err;
+        }
+    }
+
+    // TODO: this is using KM_SLEEP, is that ok?
+    let mut buf = Vec::<u8>::with_capacity(IOCTL_SZ);
+    if ddi_copyin(arg as _, buf.as_mut_ptr() as _, IOCTL_SZ, mode) != 0 {
+        return EFAULT;
+    }
+
+    let err = xde_ioc_opte_cmd(buf.as_mut_ptr() as _, arg, mode, credp, rvalp);
+
+    if ddi_copyout(buf.as_ptr() as _, arg as _, IOCTL_SZ, mode) != 0 && err == 0
+    {
+        return EFAULT;
+    }
```

Comment:
So the structure here intentionally mimics [how `dld` dispatched](
https://github.com/illumos/illumos-gate/blob/e13f92368d64a8efb300e6217cc0931bf766e6d4/usr/src/uts/common/io/dld/dld_drv.c#L1712-L1714) ioctls because I didn't want to also deal with changing that side of things at the same time.

Reply:
OK. Makes sense. Let's file a follow up issue to track cleaning this up into a way that makes more sense semantically for us (and if this is it, that's fine). It probably will be ok if we only ever do the single copyin/out, but if not it'll get a bit weirder for a consumer to understand.

Reply:
Created https://github.com/oxidecomputer/opte/issues/333 for this